src/main/cpp/src/cast_decimal_to_string.cu

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cast_string.hpp"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/convert/int_to_string.cuh>
+#include <cudf/strings/detail/converters.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/generate.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/transform.h>
+
+#include <cuda/std/climits>
+#include <cuda/std/limits>
+#include <cuda/std/type_traits>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+namespace detail {
+namespace {
+constexpr int max_zeros = 5;
+
+template <typename DecimalType>
+struct decimal_to_non_ansi_string_fn {
+  column_device_view d_decimals;
+  offset_type* d_offsets{};
+  char* d_chars{};
+
+  /**
+   * @brief Calculates the size of the string required to convert the element, in base-10 format.
+   *
+   * Output format is [-]integer.fraction
+   */
+  __device__ int32_t compute_output_size(DecimalType value)
+  {
+    auto const scale = d_decimals.type().scale();
+
+    if (scale >= 0) return strings::detail::count_digits(value) + scale;
+
+    auto const abs_value = numeric::detail::abs(value);
```

Comment:
Same comment here about overflow.

Reply:
So both of these overflows are in the kernel running now from cudf. The cudf approach doesn't verify data and lets the user do undefined things if they want and things just don't work as expected. I don't know if that is the approach we want to take or not. I'm fine either way.

Reply:
The overflow I am worried about is the min value for `DecimalType` lets take `int32_t` as an example because it is the smallest one and a scale of -1, to expose a problem. In this case `abs(-2147483648)` is `-2147483648`

This is going to mess up both the size logic and the string itself.

When calculating the size we get:

`exp_ten(1)` will be 10

Then to compute `fraction` we are doing a `count_digits` on `abs_value % exp_ten` `-2147483648 % 10` is `-8`

`count_digits` of -8 is 2, because it includes a space for the "-".

num_zeros is now 0 because `max(0, 1 - 2)` is 0

Now we calculate the output size. `value < 0` + count digits(-214748364) + 1 + 0 + 2

1 + 10 + 1 + 0 + 2 => 14


When outputting the result we get

"--214748364.-8" That is a total of 14 characters just like the count wanted, so no data overflows here, but I don't think that is the correct answer.

Reply:
To be clear I don't think this will show up in practice because the smallest decimal32 value we should ever see is `-999999999`